src/libsyntax/parse/mod.rs

```diff
@@ -271,8 +271,16 @@ pub fn char_lit(lit: &str, diag: Option<(Span, &Handler)>) -> (char, isize) {
         'u' => {
             assert_eq!(lit.as_bytes()[2], b'{');
             let idx = lit.find('}').unwrap();
-            let s = &lit[3..idx].chars().filter(|&c| c != '_').collect::<String>();
-            let v = u32::from_str_radix(&s, 16).unwrap();
+
+            // All digits and '_' are ascii, so treat each byte as a char.
+            let mut v: u32 = 0;
+            for &c in lit[3..idx].as_bytes().iter() {
+                let c = c as char;
```

Comment:
Why don't you replace that hard cast "as" with:

let c = char::from(c);

Reply:
Ok.